app/src/org/commcare/android/logic/BarcodeScanListenerDefaultImpl.java

```diff
@@ -0,0 +1,121 @@
+        if (BuildConfig.DEBUG) {
+            if (!(requestCode == BARCODE_FETCH)) {
+                throw new IllegalArgumentException("requestCode should've been BARCODE_FETCH!");
+            }
+        }
+        if (resultCode == Activity.RESULT_OK) {
+            String result = intent.getStringExtra("SCAN_RESULT");
+            barcodeScanListener.onBarcodeFetch(result, intent);
+        }
+    }
+
+    public static void onCalloutResult(BarcodeScanListener barcodeScanListener, int requestCode, int resultCode, Intent intent) {
+        if (BuildConfig.DEBUG) {
+            if (!(requestCode == CALLOUT)) {
+                throw new IllegalArgumentException("requestCode should've been CALLOUT!");
+            }
+        }
+        if (resultCode == Activity.RESULT_OK) {
+            String result = intent.getStringExtra("odk_intent_data");
+            barcodeScanListener.onCalloutResult(result, intent);
+        }
+    }
+
+    public static void callBarcodeScanIntent(Activity act) {
+        Log.i("SCAN", "Using default barcode scan");
+        Intent i = new Intent("com.google.zxing.client.android.SCAN");
+        try {
+            act.startActivityForResult(i, BARCODE_FETCH);
+        } catch (ActivityNotFoundException anfe) {
+            Toast.makeText(act,
+                    "No barcode reader available! You can install one " +
+                            "from the android market.",
+                    Toast.LENGTH_LONG).show();
+        }
+    }
+
+    public static CalloutAction makeCalloutAction(final Activity act, Callout callout, CalloutActionSetup calloutActionSetup) {
+        final CalloutData calloutData = callout.evaluate();
+
+        if (calloutData.getImage() != null && calloutActionSetup != null) {
+            calloutActionSetup.onImageFound(calloutData);
+        }
+
+        final String actionName = calloutData.getActionName();
+        final Hashtable<String, String> extras = calloutData.getExtras();
+
+        return new CalloutAction() {
+            @Override
+            public void callout() {
+                Log.i("SCAN", "Using barcode scan with action: " + actionName);
+                Intent i = new Intent(actionName);
+
+                for (Map.Entry<String, String> keyValue : extras.entrySet()) {
+                    i.putExtra(keyValue.getKey(), keyValue.getValue());
+                }
+                try {
+                    act.startActivityForResult(i, CALLOUT);
+                } catch (ActivityNotFoundException anfe) {
+                    Toast.makeText(act, "No application found for action: " + actionName, Toast.LENGTH_LONG).show();
+                }
+            }
+        };
+    }
+
+    public static View.OnClickListener makeCalloutOnClickListener(final Activity act, Callout callout, CalloutActionSetup calloutActionSetup) {
+        if (callout == null) {
```

Comment:
Whenever you have a function that checks the nullity of its arguments and performs completely separate logic based on that check, it is a sign that you should break it up into two separate functions.

In this case you should have two methods, something like: ```barcodeIntentLauncher(Activity activity)``` and ```makeCalloutWithActionStep(Activity activity, Callout callout, CalloutActionStep calloutActionStep)```.
Then ```EntitySelectActivity``` can call  ```makeCalloutWithActionStep``` and ```FormRecordList``` can call ```barcodeIntentLauncher```